evm2wast.c

```diff
@@ -0,0 +1,72 @@
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <dirent.h>
+#include <math.h>
+#include "evm2wast.h"
+#include "gadgets.h"
+#include "util.h"
+#define digits(x) (floor(log10(abs(x))) + 1) 
+/*static size_t stack_delta;
+static size_t stack_high;
+static size_t stack_low;
+static size_t gas_count;
+static char *jump_segments;
+static char *callbacks[12];
+static char *deps[256];
+*/
+static char *build_jump_map(unsigned int *segments, size_t len)
+{
+	size_t i;
+	char *brtable = calloc(len*(2+digits(len)), 1);
+	char *wasm = calloc(pow((67+2*(digits(len))), len), 1);
+	for(i=0; i<len; i++)
+	{
+		snprintf(brtable + strlen(brtable), 2+digits(len)+1, " $%ld", i);
+		snprintf(wasm + strlen(wasm), 68+(2*(digits(len))),
+		"(if(i32.eq(get_local $jump_dest)(i32.const $%d))(then (br $%ld))(else %s))",
+		segments[i], i, wasm);
+	}
+	size_t s = strlen(wasm) + 358;
+	wasm = realloc(wasm, s);
+	snprintf(wasm + strlen(wasm), s,
+	"(block $0 (if (i32.eqz (get_global $init)) (then (set_global $init (i32.const 1)) (br $0)) (else ;; the callback dest can never be in the first block (if (i32.eq (get_global $cb_dest) (i32.const 0)) (then %s) (else ;; return callback destination and zero out $cb_dest get_global $cb_dest (set_global $cb_dest (i32.const 0)) (br_table $0 ${brTable}))))))", wasm);
+	return wasm;
+}
+static char *assemble_segments(unsigned int *segments, size_t len)
+{
+	char *wasm = build_jump_map(segments, len);
+	char *ret = malloc(len*(8+digits(len)+strlen(wasm)) + 122);
+	strcat(ret, "(func $main (export \"main\") (local $jump_dest i32) (set_local $jump_dest (i32.const -1)) (block $done (loop $loop ");
+	unsigned int i;
+	for(i=0; i<len; i++)
+	{
+		sprintf(ret+strlen(ret), "(block $%d %s", i+1, wasm);
+	}
+	return ret;
+}
+int evm2wasm(char *evm_code, size_t len, char *wast_code, size_t wast_size)
```

Comment:
This should be called `evm2wast`. Also I think `wast_size` should be a pointer updated with the final length.

Reply:
`wast_code` must be a pointer of a pointer if you want to use `realloc` below